Initial version of parallelization framework

chenccfb opened the pull request:
Summary:
There are two parallelization mode: data and model parallelization.

This diff only implements data parallelization for FC node, i.e., torch.linear.

I will have a follow up diff to implement parallelization for other ops.

## Model Parallelization:

        x   param                       x   param                           x   param
         \ /                             \ /                                  \ /
         add    weight  bias     ====>    add   weight_1  bias_1    ...        add(reused)   weight_4  bias_4
           \      |      /                 \       |       /                     \             |       /
            \     |     /                   \      |      /                       \            /      /
                linear                           linear1            ...                linear4
                  |                                   \             |                   /
                clamp                                  \            |                  /
                                                                concat_linear
                                                                    |
                                                                  clamp
                                                                    |
                                                                  output

## Data Parallelization:

      x   param                           x   param
         \ /                                  \ /
         add                                  add
          |                 ==>         /    /   |     \
        linear                         /    |    |      \
          |                      slice1 slice2 slice3 slice4
         clamp                        |     |    |        |
          |                     linear1  linear2 linear3 linear4
        output                         \    |     |       /
                                         \  |     |      /
                                          concat_linear
                                               |
                                             clamp
                                               |
                                             output

Test Plan:
## Unit Tests
buck test glow/fb/fx/fx_glow:test_fx_parallelize

Differential Revision: D26812645

jamesr66a commented on torch/fx/node.py:
```diff
@@ -439,6 +439,20 @@ def is_impure(self):
 
         return False
 
+    def replace_input_with(self, old_input: 'Node', new_input: 'Noe'):
```
s/'Noe'/'Node'/

chenccfb replied:
Good catch!

jamesr66a commented on torch/fx/node.py:
```diff
@@ -439,6 +439,20 @@ def is_impure(self):
+    def replace_input_with(self, old_input: 'Node', new_input: 'Noe'):
+        """
+        Replace input node with new input node.
+        """
```
Please format docstrings according to the [google docstring style guide](https://google.github.io/styleguide/pyguide.html), including a description, `Args`, `Return`, and ideally an example invocation

jamesr66a commented on torch/fx/node.py:
```diff
@@ -439,6 +439,20 @@ def is_impure(self):
+        new_kwargs = map_arg(self.kwargs, maybe_replace_node)
+        assert isinstance(new_args, tuple)
+        assert isinstance(new_kwargs, dict)
+        self.__update_args_kwargs(new_args, new_kwargs)
```
Overall I'm fine with having this method, however I'd like to understand the UX in terms of discoverability of the `map_arg` pattern. How hard is it to know to use `map_arg`? Do you think that should be a method on `Node` that applies to both `args` and `kwargs`? (essentially a generalization of this function)

chenccfb replied:
> How hard is it to know to use map_arg

I think ```map_arg``` is pretty easy to use and powerful and I am happy with that :)

chenccfb committed:
Initial version of parallelization framework (#53915)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/53915

Differential Revision: D26812645

fbshipit-source-id: d589b0df2ff50186b0e53905822cfe30f73bcecf
